### PR TITLE
test_state_transfer wouldn't fail correctly for f>1

### PR DESCRIPTION
**Pull request by KTeofanidis**
Due to only stopping one node the test would always succeed irrespective of stale_node working correctly or not when f>1

### CODE REVIEW

File: tests/apollo/test_skvbc.py [hunk 1]
```diff
@@ -75,9 +75,8 @@ async def test_state_transfer(self, bft_network,exchange_keys=True):
         await bft_network.wait_for_state_transfer_to_start()
         await bft_network.wait_for_state_transfer_to_stop(0, stale_node)
         await skvbc.assert_successful_put_get(self)
-        random_replica = random.choice(
-            bft_network.all_replicas(without={0, stale_node}))
-        bft_network.stop_replica(random_replica)
+        bft_network.stop_replicas(bft_network.random_set_of_replicas(
+            bft_network.config.f, without={0, stale_node}))
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc.py:79]: This is great, but can you actually move this logic into the `bft.force_quorum_including_replica()` method and use it wherever applicable?

**Review by teoparvanov — COMMENTED** [review]
Great find, thanks @KTeofanidis ! I have a suggestion how to make this even better...

File: tests/apollo/util/bft.py [hunk 2]
```diff
@@ -630,18 +630,9 @@ def force_quorum_including_replica(self, replica_id, primary=0):
...
-                action.log(f'Stopping backup replica {backup_replica_id} in order '
-                      f'to force a quorum including replica {replica_id}...')
-                self.stop_replica(backup_replica_id)
-                if len(self.procs) == 2 * self.config.f + self.config.c + 1:
-                    break
-
-            assert len(self.procs) == 2 * self.config.f + self.config.c + 1
+            assert len(self.procs) >= 2 * self.config.f + self.config.c + 1
+            self.stop_replicas(self.random_set_of_replicas(
+                len(self.procs) - (2 * self.config.f + self.config.c + 1), without={primary, replica_id}))
```
Comment (teoparvanov) [inline tests/apollo/util/bft.py:635]: why not just 
```python
            self.stop_replicas(self.random_set_of_replicas(f, without={primary, replica_id}))
```
?

Comment (KTeofanidis) [reply tests/apollo/util/bft.py:635]: The original code was correct. We need at least 2f + c + 1 nodes for the quorum, so for example in the edge case n6 c1 f1 we need to stop 2 instead of 1 to ensure the relevant replica_id is a non failed part of the quorum of 4.

Comment (teoparvanov) [reply tests/apollo/util/bft.py:635]: ok, makes sense

File: tests/apollo/test_skvbc.py [hunk 3]
```diff
@@ -75,9 +75,7 @@ async def test_state_transfer(self, bft_network,exchange_keys=True):
         await bft_network.wait_for_state_transfer_to_start()
         await bft_network.wait_for_state_transfer_to_stop(0, stale_node)
         await skvbc.assert_successful_put_get(self)
-        random_replica = random.choice(
-            bft_network.all_replicas(without={0, stale_node}))
-        bft_network.stop_replica(random_replica)
+        bft_network.force_quorum_including_replica(stale_node)
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc.py:78]: can you also use this function in the test you recently added in the network partitioning suite?

**Review by teoparvanov — APPROVED** [review]
LGTM, thanks!

File: tests/apollo/util/bft.py (hunk 2 above)
Comment (teoparvanov) [inline tests/apollo/util/bft.py:635]: Passing the primary as a parameter could lead to inconsistencies. Could you instead dynamically retrieve it, such as:
```python
primary = await bft_network.get_current_primary()
```
and also remove it from the method's parameters?

**Review by teoparvanov — CHANGES_REQUESTED** [review]
one last thing before we merge this - see my comment above

File: tests/apollo/util/bft.py (hunk 2 above)
Comment (KTeofanidis) [reply tests/apollo/util/bft.py:635]: This is going to add some overhead considering currently the method isn't async. Are you sure?

Comment (teoparvanov) [reply tests/apollo/util/bft.py:635]: Yes

Comment (teoparvanov) [reply tests/apollo/util/bft.py:635]: PS: don't forget to add `await` when calling the force quorum method

**Review by teoparvanov — APPROVED** [review]
Looks great, thanks @KTeofanidis !
